### PR TITLE
Avail DA adapter Integration into rollkit

### PR DESCRIPTION
**Pull request by chandiniv1**
This PR consist DA adapter for the [Avail](http://availproject.org/).

## Summary by CodeRabbit

- **New Features**
	- Introduced functionality for interacting with an availability layer for data storage.
	- Launched a gRPC service for the availability layer.
	- Added a mock server for handling block data-related HTTP requests.

- **Documentation**
	- Provided comprehensive setup and usage instructions for the Avail DA package.

- **Tests**
	- Integrated actual test execution commands in the continuous integration workflow.
	- Separated unit test execution from other test types in the Makefile.

- **Chores**
	- Updated the Makefile to run a complete suite of tests.

### CODE REVIEW

**Comment by nashqueue** [comment]
Thank you ! Can you make sure that the CI passes ?

**Comment by chandiniv1** [comment]
> Thank you ! Can you make sure that the CI passes ?

@nashqueue I have fixed the CI passes

File: avail_test.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package avail_test
+
+import (
+	"testing"
+)
+
+func TestAvailDA(t *testing.T) {
+	t.Skip()
```
Comment (Manav-Aggarwal) [inline avail_test.go:8]: Please add relevant unit tests to ensure code robustness

**Review by Manav-Aggarwal — CHANGES_REQUESTED** [review]
Could you explain how this code has been tested, specifically the `Submit` and `Get` methods? Otherwise, looks great, just missing unit tests.

**Comment by chandiniv1** [comment]
> Could you please explain how this code has been tested, specifically the `Submit` and `Get` methods? Otherwise, looks great, just missing unit tests.
> 
> The CI currently does not test any core functionality except linting checks. See [CI](https://github.com/rollkit/avail-da/blob/f59576ddaa86f61884d905fe3125382aeaa2c588/.github/workflows/test.yml#L29)

@Manav-Aggarwal apart from writing unit tests, we haven't tested the application since the current version of rollkit is not compatible with GM application.

File: go.mod [hunk 2]
```diff
@@ -1,3 +1,20 @@
 module github.com/rollkit/avail-da
 
-go 1.21.0
+go 1.21.1
```
Comment (MSevey) [inline go.mod:3]: We should add a README with this PR explaining how to use the interface, including the required go version.

**Review by MSevey — COMMENTED** [review]
lgtm, minor comment about adding README and will re-review once unit tests are added.

**Comment by chandiniv1** [comment]
@Manav-Aggarwal @MSevey  I have added the readme regarding the usage of the da interface, Please have a look

**Comment by chandiniv1** [comment]
> > @Manav-Aggarwal @MSevey I have added the readme regarding the usage of the da interface, Please have a look
> 
> @chandiniv1 thanks for adding unit tests! Can you enable the tests by removing the echo command and uncommenting the code here:
> 
> https://github.com/rollkit/avail-da/blob/f59576ddaa86f61884d905fe3125382aeaa2c588/.github/workflows/test.yml#L29
> 
> Also I don't see the README, did it maybe not get committed?
> 
> Lastly, looks like golangci-lint is failing, can you address that please?

@MSevey I have enabled tests and fixed lint errors, please have a look

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -50,4 +49,4 @@ jobs:
         with:
           go-version: ${{ inputs.GO_VERSION }}
       - name: Integration Tests
-        run: echo "No integration tests yet"
+        run: make test
```
Comment (MSevey) [inline .github/workflows/test.yml:52]:
```suggestion
        run: echo "No integration tests yet"
```

**Review by MSevey — APPROVED** [review]
Great work!  Thanks!

File: avail.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	var blockNumber uint32
+	for _, id := range ids {
+	Loop:
+		blockNumber = binary.BigEndian.Uint32(id)
+		blocksURL := fmt.Sprintf(c.config.LcURL+BlockURL, blockNumber)
+		parsedURL, err := url.Parse(blocksURL)
+		if err != nil {
+			return nil, err
+		}
+		time.Sleep(10 * time.Second)
```
Comment (tuxcanfly) [inline avail.go:187]: This sleeps in the first iteration as well, right? Might be better to move this to the `else ... == PROCESSING_BLOCK` condition.

File: go.mod [hunk 5]
```diff
@@ -1,3 +1,26 @@
...
+
+toolchain go1.21.5
+
+require (
+	github.com/rollkit/go-da v0.0.0-20231024133951-57bc36006772
```
Comment (tuxcanfly) [inline go.mod:8]: go-da latest `v0.2.0` adds context and gas price for submit. Something to note for later.

File: go.mod [hunk 6]
```diff
@@ -1,3 +1,26 @@
...
+	github.com/pmezard/go-difflib v1.0.0 // indirect
+	golang.org/x/net v0.19.0 // indirect
+	golang.org/x/sys v0.15.0 // indirect
+	golang.org/x/text v0.14.0 // indirect
+	google.golang.org/genproto/googleapis/rpc v0.0.0-20230822172742-b8732ec3820d // indirect
+	google.golang.org/protobuf v1.31.0 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
+)
+
+replace github.com/rollkit/go-da => github.com/rollkit/go-da v0.0.0-20231207150926-93600f28d67d
```
Comment (tuxcanfly) [inline go.mod:26]: This seems unnecessary, just use the commit in the require clause.

**Review by tuxcanfly — DISMISSED** [review]
Good work! Added some suggestions that can go in the next round.

**Review by Manav-Aggarwal — DISMISSED** [review]
proxy utACK

**Review by MSevey — APPROVED** [review]

**Review by Manav-Aggarwal — APPROVED** [review]

**Pull request merged**